Consistent path attribute for file-related modules

dagwieers opened the pull request:
##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
acl, blockinfile, ini_file, lineinfile, mount, replace, stat, win_lineinfile, xattr

##### ANSIBLE VERSION
<!--- Paste verbatim output from “ansible --version” between quotes below -->
v2.2

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->
Not all file-related modules consistently use "path" as the attribute to specify a single filename, some use "dest", others use "name". Most do have aliases for either "name" or "destfile".

This change makes "path" the default attribute for (single) file-related modules, but also adds "dest" and "name" as aliases, so that people can use a consistent way of attributing paths, but also to ensure backward compatibility with existing playbooks.

This is related to an older PR ansible/ansible-modules-core#5437

dagwieers committed:
Consistent path attribute for file-related modules

Not all file-related modules consistently use "path" as the attribute to specify a single filename, some use "dest", others use "name". Most do have aliases for either "name" or "destfile".

This change makes "path" the default attribute for (single) file-related modules, but also adds "dest" and "name" as aliases, so that people can use a consistent way of attributing paths, but also to ensure backward compatibility with existing playbooks.

NOTE: The reason for changing this, is that it makes Ansible needlessly harder to use if you have to remember that e.g. the xattr module requires the name attribute, the lineinfile module requires a dest attribute, and the stat module requires a path attribute.

dagwieers commented:
The reason for me for changing this, is that it makes Ansible needlessly harder to use if you have to remember that e.g. the **xattr** module requires the **name** attribute, the **lineinfile** module requires a **dest** attribute, and the **stat** module requires a **path** attribute.

bcoca commented:
I've been meaning to do this for a long time, just could not get everyone to agree on which 'name' should be the default ....

dagwieers commented:
I am open to change the default (per module if there's no consensus), but at least now it also works for the various standards implemented (dest, name, path) and I think that is the more important part of this change.

PS The Shippable error is due to an HTTP timeout error unrelated to this change. (Shippable seems to fail more than work these days :-/)

bcoca commented:
I noted the shippable issue, it happens a lot when we depend on external resources or race conditions exist and their system is loaded, ignoring that for this PR.

As for final term, I added this to agenda for discussion today in core meeting.

dagwieers commented:
On another PR it was the docker infrastructure that timed out (on a Fedora env).
I fear that if we don't record these issues, they go unnoticed and may not be investigated (in bulk).
Not sure if these failures are automatically escalated/logged for troubleshooting since it may not be our infrastructure that fails. But having these kind of statistics may help a lot to spend effort where it is most required.

@bcoca Thanks for the due diligence !

gundalow commented on lib/ansible/modules/files/blockinfile.py:
```diff
@@ -104,28 +102,32 @@
     version_added: "2.1"
+notes:
+  - This module supports check mode.
+  - When using 'with_*' loops be aware that if you do not set a unique mark the block will be overwritten on each iteration.
```
Did you meant to include this?

dagwieers replied:
Yes, it was actually moved to the same location as the other modules (below the options).
At first I added the section, and then a test failed because `notes:` was twice in the dictionary :-(

bcoca commented on lib/ansible/modules/files/acl.py:
```diff
@@ -27,12 +27,13 @@
     required: true
     default: null
     description:
       - The full path of the file or object.
```
description should have something like: Before 2.3 this option was only usable as 'name'

dagwieers replied:
Ah, you want that as part of the description ? I added it as a note.

The reason I did not want to put that into the description itself, is that in the documentation you can see the aliases straight away. So it is already obvious both work fine.

dagwieers replied:
Hmm, ok, if the documentation is changed right now that will be moot. Will add.

dagwieers replied:
In this case both name and path used to work, they just switched cases.
But in other cases you are right.

dagwieers replied:
Not sure if this makes sense for the **blockinline** module:

```yaml
options:            
  path:
    aliases: [ dest, destfile, name ]
    required: true
    description:
      - The file to modify.
        Before 2.3 this option was only usable as 'dest', 'destfile' and 'name'  
```

dagwieers replied:
Implemented.

bcoca commented on lib/ansible/modules/files/acl.py:
```diff
@@ -27,12 +27,13 @@
       - The full path of the file or object.
-    aliases: ['path']
+    aliases: ['name']
+    version_added: 'historical'
```
dont add `version_added: historical`

dagwieers replied:
ok, but all tests will fail.

bcoca commented on lib/ansible/modules/files/acl.py:
```diff
@@ -97,27 +97,28 @@
 '''
 
 EXAMPLES = '''
 # Grant user Joe read access to a file
```
another note about the change and version in examples would be nice

dagwieers replied:
As a comment, you mean ?

dagwieers replied:
Implemented.

bcoca commented on lib/ansible/modules/files/blockinfile.py:
```diff
@@ -194,7 +196,7 @@ def check_file_attrs(module, changed, message):
     module = AnsibleModule(
         argument_spec=dict(
-            dest=dict(required=True, aliases=['name', 'destfile'], type='path'),
+            path=dict(required=True, aliases=['dest', 'destfile', 'name'], type='path'),
```
~~i dont see the need for changing the internal variable names~~ this is argspec, it is needed here

dagwieers commented:
Alright, I implemented everything that was requested.
- Add 'path' and add the previous default as alias only
- Add a note to the path option description, to indicate that this has changed with v2.3
- Add a note to the module, to indicate that this has changed with v2.3
- Add a comment to the examples, to indicate that this has changed with v2.3
- Removed to `version_added: 'historical`

dagwieers commented:
shipit

jctanner commented:
!needs_revision

dagwieers commented:
Any more feedback ?

gundalow reviewed (approved):
shipit

bcoca commented:
i keep refreshing, but i still dont see the updates to description

dagwieers commented:
@bcoca I think everything is implemented what we discussed, if you don't agree, tell me what is missing and I will add it.
